newIDE/app/src/Profile/CreateAccountDialog.js

```diff
@@ -76,119 +73,137 @@ export const getPasswordErrorText = (error: ?AuthError) => {
   return undefined;
 };
 
-export default class CreateAccountDialog extends Component<Props, State> {
-  state = {
-    form: {
-      email: '',
-      password: '',
-      username: '',
-    },
-  };
-
-  _canCreateAccount = () => {
-    return (
-      !this.props.createAccountInProgress &&
-      isUsernameValid(this.state.form.username, true)
-    );
-  };
-
-  _onCreateAccount = () => {
-    if (!this._canCreateAccount()) return;
+const CreateAccountDialog = ({
+  onClose,
+  onGoToLogin,
+  onCreateAccount,
+  createAccountInProgress,
+  error,
+}: Props) => {
+  const [email, setEmail] = React.useState<string>('');
+  const [password, setPassword] = React.useState<string>('');
+  const [username, setUsername] = React.useState<string>('');
+  const [
+    usernameAvailability,
+    setUsernameAvailability,
+  ] = React.useState<?UsernameAvailability>(null);
+  const [
+    isValidatingUsername,
+    setIsValidatingUsername,
+  ] = React.useState<boolean>(false);
 
-    const { form } = this.state;
-    this.props.onCreateAccount(form);
-  };
-
-  render() {
-    const { onClose, createAccountInProgress, onGoToLogin, error } = this.props;
+  const canCreateAccount = React.useMemo(
+    () =>
+      !createAccountInProgress &&
+      isUsernameValid(username, { allowEmpty: true }) &&
+      !isValidatingUsername &&
+      (!usernameAvailability || usernameAvailability.isAvailable),
+    [
+      username,
+      createAccountInProgress,
+      isValidatingUsername,
+      usernameAvailability,
+    ]
+  );
+  const createAccount = React.useCallback(
```

Comment:
I don't know if it's worth using a callback here since every dependency will change a lot. I saw articles saying that in that case, using callbacks could impact performance unnecessarily

Reply:
yep, I'll remove that. 
It's definitely a bad practice to use them everywhere